src/main/java/com/shapesecurity/salvation/Parser.java

```diff
@@ -351,12 +353,19 @@ private void enforceMissingDirectiveValue(@Nonnull Token directiveNameToken) thr
     @Nonnull private Set<SourceExpression> parseSourceList() throws DirectiveParseException {
         Set<SourceExpression> sourceExpressions = new LinkedHashSet<>();
         boolean parseException = false;
-        boolean seenNone = false;
+        Map<String, Boolean> seenStates = new HashMap<>();
```

Comment:
Instead of using booleans, we can just use existence, which means we should use a set.